### PR TITLE
Port COCOA to rely on DD4hep

### PR DESCRIPTION
**Pull request by ghugo83**
Moved COCOA DD dependencies to DD4hep.
As discussed in https://github.com/cms-sw/cmssw/pull/25356. 

**General description of COCOA:**
https://cds.cern.ch/record/1047121/files/p193.pdf
COCOA is now basically standalone alignment software. 
It is designed to simulate and reconstruct optical alignment systems for CMS. It was also used (in the past I presume) to analyze the CMS optical alignment data, and pass the alignment correction to the CMS reconstruction.

**PR description**:
Following was done for port to DD4hep (this can also be seen as a HOWTO):
- DetectorDescription/Core -> DetectorDescription/DDCMS dependencies.
- Switch to DD4hep-based versions of DDCompactView and DDFilteredView.
- Switch to DD4hep-based geometry building of a cms::DDDetector optical alignment system, with DDDetectorESProducer.
- Remove use of old DDExpandedNode and navigation.
- DDLogicalPart -> dd4hep::PlacedVolume and her friends.
- DDTranslation -> dd4hep::Direction (ROOT::Math::XYZVector)
- DDRotationMatrix -> dd4hep::Rotation3D (ROOT::Math::Rotation3D)
- Remove use of DDsvalues_type and DDValue -> Instead, use of cms::DDSpecParRegistry,  cms::DDSpecParRefs, and cms::DDSpecPar::value<T>.

**PR validation:**
Had to 'reconstruct' input files to be able to run COCOA.
Notably created an input XML file + an OpticalAlignmentsRcd stored in local .db.
Created small guide: Alignment/CocoaApplication/README.md, so that one can now easily run COCOA and reproduce results.
No regression due to port to DD4hep: 
- Fully checked ideal geometry building, perfectly matching geometry from XML.
- Checked parameters and measurements from XMLs. 

**Bugs in COCOA:**
While checking for regressions, I spotted & fixed the following issues in COCOA (independent from port to DD4hep). Visible in the code before PR:
- https://github.com/cms-sw/cmssw/blob/CMSSW_10_3_X/Alignment/CocoaApplication/src/CocoaAnalyzer.cc#L214
colX, colY and colZ are obviously not columns, but rows. 
This results in the rotation matrix rotclhep to be ill-formed (transpose, hence inverse, of what it should be). Hence, the local rotations in ideal geometry were inverse of what they should be, and could not be properly checked against those of the geometry from DB.
- https://github.com/cms-sw/cmssw/blob/CMSSW_10_3_X/Alignment/CocoaFit/src/CocoaDBMgr.cc#L191
This is used to write results to DB. Error in the conversion (!!!), should go from m (in COCOA) to cm (in DB). Hence data stored in DB was wrong...
- https://github.com/cms-sw/cmssw/blob/CMSSW_10_3_X/Alignment/CocoaApplication/src/CocoaAnalyzer.cc#L343 
std::string type is initialized with value from previous iteration. Hence, within the loop, there is a telescopage of its values. This leads to dimFactor being not properly calculated, hence all 'extra alignment parameters' values were wrong.

**Disclaimer:**
The COCOA packages were very bad style, I re-wrote and commented only part of them, in the context of the port to DD4hep. Notably, src/CocoaAnalyzer.cc was removed then restored, hence I am partially attributed parts of old code there which is not mine.
I fully debugged COCOA behavior in the fields of ideal geometry building + measurements corrections + results write to DB. Though, there are other important aspects in the alignment which I did not debug (notably the least-square parameters fitting).

### CODE REVIEW

**Comment by ghugo83** [comment]
Adding @ianna @civanch @cvuosalo

**Comment by civanch** [comment]
please test

**Review by ianna — CHANGES_REQUESTED** [review]
@ghugo83 - Good job! Some comments are just suggestions. Using `MessageLogger` is important and please, squash the commits if you can - it would be easier to scroll to a relevant discussion. Thanks.

File: Alignment/CocoaApplication/src/CocoaAnalyzer.cc [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+/* Helper: For a given node, get the value associated to a given parameter, from the XMLs SpecPar sections.
+ * This is the parameterValueIndex-th value (within a XML SpecPar block.) of the desired parameter. 
+ */
+template <typename T>
+T CocoaAnalyzer::getParameterValueFromSpecParSections(const cms::DDSpecParRegistry& allSpecParSections,
+                                                      const std::string& nodePath,
+                                                      const std::string& parameterName,
+                                                      const unsigned int parameterValueIndex) {
+  const std::vector<T>& allParameterValues =
+      getAllParameterValuesFromSpecParSections<T>(allSpecParSections, nodePath, parameterName);
```
[the remark this answers is not in the document]
Comment (ghugo83) [reply Alignment/CocoaApplication/src/CocoaAnalyzer.cc:609]: Thanks @ianna  for the review!
Yes the thing is that apart from the README and getAllParameterValuesFromSpecParSections<T>, all the other comments address (very old) code which is not mine, it was already like that in 10_3_X:
https://github.com/cms-sw/cmssw/blob/CMSSW_10_3_X/Alignment/CocoaApplication/interface/CocoaAnalyzer.h 
https://github.com/cms-sw/cmssw/blob/CMSSW_10_3_X/Alignment/CocoaApplication/src/CocoaAnalyzer.cc
When CocoaAnalyzer is introduced back after https://github.com/cms-sw/cmssw/pull/25356 , it also contains the sections which were not modified in the context of the to port to DD4hep.
But ok, will address it :)

File: Alignment/CocoaApplication/src/CocoaAnalyzer.cc [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+    fileout << "............ program ended OK" << std::endl;
+  }
+}
+
+/* Helper: For a given node, get the values associated to a given parameter, from the XMLs SpecPar sections.
+ * NB: The same parameter can appear several times WITHIN the same SpecPar section (hence, we have a std::vector).
+ * WARNING: This stops at the first relevant SpecPar section encountered.
+ * Hence, if A GIVEN NODE HAS SEVERAL SPECPAR XML SECTIONS RE-DEFINING THE SAME PARAMETER,
+ * only the first XML SpecPar block will be considered.
+ */
```
[the remark this answers is not in the document]
Comment (ghugo83) [reply Alignment/CocoaApplication/src/CocoaAnalyzer.cc:576]: Hmm within the same SpecPar block, all parameters values are stored in a std::vector.
But a cms::DDSpecParRefs itself is a std::vector<const DDSpecPar*>, ie a vector of SpecPar blocks.

File: Alignment/CocoaApplication/src/CocoaAnalyzer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (ghugo83) [reply Alignment/CocoaApplication/src/CocoaAnalyzer.cc:601]: @ianna Ok I addressed the review :)
Regarding the commits: is it possible to just do a squash-and-merge when merging the PR?

[the remark this answers is not in the document]
Comment (ianna) [reply Alignment/CocoaApplication/src/CocoaAnalyzer.cc:601]: @ghugo83 - Thanks! Here is the link to CMS FAQ http://cms-sw.github.io/faq.html#how-do-i-collapse-multiple-commits-into-one 
Note, it's considered an advanced user question :-)

[the remark this answers is not in the document]
Comment (ghugo83) [reply Alignment/CocoaApplication/src/CocoaAnalyzer.cc:601]: @ianna There are repos with a github feature called 'Squah and merge' allowed. But dont know if that is the case for CMSSW. 
Ok but usually when I do git --amend, I cannot push to the remote github branch anymore, without doing a forced push. Will it still be fine, given an open PR?

[the remark this answers is not in the document]
Comment (ghugo83) [reply Alignment/CocoaApplication/src/CocoaAnalyzer.cc:601]: Ok all my work is now in the second commit :)

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by christopheralanwest** [comment]
assign geometry

Since this PR involves DD4hep, I would like to have a sign-off from geometry prior to signing off for alca.

**Comment by cvuosalo** [comment]
+1

**Comment by christopheralanwest** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**